upickle/implicits/src-3/upickle/implicits/macros.scala

```diff
@@ -199,6 +199,63 @@ def tagKeyImpl[T](using Quotes, Type[T])(thisOuter: Expr[upickle.core.Types with
     case None => '{${thisOuter}.tagName}
   }
 
+inline def applyConstructor[T](params: Array[Any]): T = ${ applyConstructorImpl[T]('params) }
+def applyConstructorImpl[T](using quotes: Quotes, t0: Type[T])(params: Expr[Array[Any]]): Expr[T] =
+  import quotes.reflect._
+  def apply(typeApply: Option[List[TypeRepr]]) = {
+    val tpe = TypeRepr.of[T]
+    val companion: Symbol = tpe.classSymbol.get.companionModule
+    val constructorParamSymss = tpe.typeSymbol.primaryConstructor.paramSymss
+
+    def isPrimaryApplyMethod(syms1: Seq[Seq[Symbol]], syms2: Seq[Seq[Symbol]]) = {
+      // try to guess the primary apply method based on the parameter counts
+      // not sure why comparing the types doesn't seem to work
+      // println(syms1.flatten.zip(syms2.flatten).map{case (s1, s2) => (s1.typeRef.simplified, s2.typeRef.simplified)})
```

Comment:
@bishabosha do you have any idea how I can find the `apply` method with a signature matching the primary constructor? In Scala 2 I can just call `apply` in a quasiquote and it somehow resolves to the right now, but in Scala 3 it seems pickier and asks me to select which apply method I want to call ahead of time

Reply:
I think you would still want to capture the `MirroredElemTypes` from the mirror - cast each element of the array to the type from `MirroredElemTypes`, and then call `Select.overloaded(ref, "apply", args)` - however then that makes it hard to adapt the varargs if needed - I wonder if its not needed

Reply:
or yeah you maybe can manually resolve the right apply method by comparing the types of the mirrored elem types

Reply:
I don't think you need the mirrored elem types - cast away with the primary constructor signatures

Reply:
How do I compare the primary constructor signatures though? `println(syms1.flatten.zip(syms2.flatten).map{case (s1, s2) => (s1.typeRef, s2.typeRef)})` seems to give me singleton `TypeRef`s or something for each of the parameters, and when I try to `=:=` them I get `false`. Also, if the case class i generic, then presumably the `T` type param for the primary constructor is different from the `T` type param from the `apply` method

Reply:
there's the substituteTypes method, where you can substitute references to the constructor type parameters, with the type arguments of the typeclass

Reply:
I added a suggestion that passes the tests